File: bindings/python/chonkdock/wire.py
"""The chonkstep dockapp wire codec, protocol version 1.

A line-for-line transcription of ``crates/chonk-dock-proto/src/wire.rs``
into stdlib Python. Everything here is pure: bytes in, messages out, no
I/O. The transport is ``SOCK_SEQPACKET``, so there is no framing to
implement — one ``send()`` is one message, and the byte layouts below
start with the universal 4-byte header (``kind`` + three zero bytes).

Decoding is strict on purpose, mirroring the reference decoder: unknown
kinds, non-zero reserved bytes, trailing bytes, out-of-range enums,
over-cap strings and unusable floats are all rejected with
``DecodeError`` rather than clamped or ignored. The shell is the
trusted end of a dockapp's socket, but "trusted" is a statement about
intent, not about bugs — a client that silently mis-parses a message
draws garbage with no error anywhere to explain it.

See ``docs/dockapp-protocol.md`` for the full contract in prose.
"""

from __future__ import annotations

import struct
from dataclasses import dataclass

PROTOCOL_VERSION = 1
TOKEN_BYTES = 16
MAX_MESSAGE_BYTES = 256 * 1024
MAX_FRAME_BYTES = MAX_MESSAGE_BYTES - 64
MAX_TILE_PX = 256
MAX_SCALE = 8.0
MAX_TILE_UNITS = 4
MAX_ID_BYTES = 64
MAX_LOG_BYTES = 256
MAX_THEME_ID_BYTES = 64
MAX_THEME_TOML_BYTES = 128 * 1024

# Message kinds. Client->shell in the low space, shell->client with the
# high bit set, so a reflected message is an UnknownKind and never a
# reinterpretation.
KIND_HELLO = 0x01
KIND_FRAME = 0x02
KIND_PONG = 0x03
KIND_LOG = 0x04
KIND_WELCOME = 0x81
KIND_THEME_CHANGED = 0x82
KIND_INPUT = 0x83
KIND_VISIBILITY = 0x84
KIND_PING = 0x85
KIND_GOODBYE = 0x86

# Input mask bits for Hello's `wants` field.
WANT_PRESS = 1 << 0
WANT_RELEASE = 1 << 1
WANT_SCROLL = 1 << 2
WANT_CROSSING = 1 << 3
WANT_ALL = WANT_PRESS | WANT_RELEASE | WANT_SCROLL | WANT_CROSSING

# InputEvent.kind
INPUT_PRESS = 1
INPUT_RELEASE = 2
INPUT_SCROLL = 3
INPUT_ENTER = 4
INPUT_LEAVE = 5

# InputEvent.button (0 = none)
BUTTON_LEFT = 1
BUTTON_MIDDLE = 2  # never actually sent by the shell
BUTTON_RIGHT = 3  # never actually sent by the shell

# Log levels
LOG_ERROR = 1
LOG_WARN = 2
LOG_INFO = 3
LOG_DEBUG = 4

# Goodbye reasons
GOODBYE_SHUTDOWN = 1
GOODBYE_PROTOCOL_ERROR = 2
GOODBYE_UNAUTHORIZED = 3
GOODBYE_REPLACED = 4
GOODBYE_TILE_TOO_LARGE = 5
GOODBYE_OVERFLOW = 6
GOODBYE_REMOVED = 7

GOODBYE_NAMES = {
    GOODBYE_SHUTDOWN: "Shutdown",
    GOODBYE_PROTOCOL_ERROR: "ProtocolError",
    GOODBYE_UNAUTHORIZED: "Unauthorized",
    GOODBYE_REPLACED: "Replaced",
    GOODBYE_TILE_TOO_LARGE: "TileTooLarge",
    GOODBYE_OVERFLOW: "Overflow",
    GOODBYE_REMOVED: "Removed",
}

_ID_CHARS = frozenset(
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_.:"
)


class DecodeError(ValueError):
    """The peer's bytes could not be read as a v1 message."""


class EncodeError(ValueError):
    """Local code tried to build a message that cannot legally exist."""


def is_valid_id(dockapp_id: str) -> bool:
    """The wire's id rule: 1..=64 bytes of ``[A-Za-z0-9._:-]``."""
    return (
        0 < len(dockapp_id) <= MAX_ID_BYTES
        and all(c in _ID_CHARS for c in dockapp_id)
    )


def frame_fits(tile_px: int, tile_units: int) -> bool:
    """Whether a tile of this geometry can cross the v1 inline transport."""
    if not (0 < tile_px <= MAX_TILE_PX and 0 < tile_units <= MAX_TILE_UNITS):
        return False
    return tile_px * tile_px * tile_units * 4 <= MAX_FRAME_BYTES


@dataclass(frozen=True)
class ThemeState:
    """The geometry and palette a dockapp draws with.

    ``theme_id`` is the fast path (look the palette up if you ship
    them); ``theme_toml`` is the correctness path (a serialized theme
    table, possibly empty). A client that can use neither falls back to
    its own default colors — wrong colors, never a blank tile.
    """

    tile_px: int
    scale: float
    theme_id: str
    theme_toml: str

    def height_for(self, tile_units: int) -> int:
        """The frame height for a dockapp of `tile_units` stacked tiles."""
        return self.tile_px * tile_units


@dataclass(frozen=True)
class InputEvent:
    """One pointer event, in coordinates local to this dockapp's tile."""

    kind: int  # INPUT_*
    button: int  # BUTTON_* or 0
    x: int
    y: int
    delta: int  # scroll notches; 0 for everything but INPUT_SCROLL


def _header(kind: int) -> bytes:
    return struct.pack("<B3x", kind)


def _check_header(buf: bytes) -> int:
    if len(buf) == 0:
        raise DecodeError("empty message")
    if len(buf) > MAX_MESSAGE_BYTES:
        raise DecodeError(f"message of {len(buf)} bytes exceeds the cap")
    if len(buf) < 4:
        raise DecodeError("message ended inside the header")
    if buf[1:4] != b"\x00\x00\x00":
        raise DecodeError("reserved header bytes were not zero")
    return buf[0]


def _exact(buf: bytes, want: int, what: str) -> None:
    if len(buf) < want:
        raise DecodeError(f"message ended inside {what}")
    if len(buf) > want:
        raise DecodeError(f"{len(buf) - want} trailing bytes after {what}")


# ---------------------------------------------------------------------
# Client -> shell encoders
# ---------------------------------------------------------------------

def encode_hello(dockapp_id: str, tile_units: int, token: bytes,
                 wants: int = WANT_ALL) -> bytes:
    if not is_valid_id(dockapp_id):
        raise EncodeError(f"invalid dockapp id {dockapp_id!r}")
    if len(token) != TOKEN_BYTES:
        raise EncodeError(f"token must be {TOKEN_BYTES} bytes")
    if wants & ~WANT_ALL:
        raise EncodeError(f"undefined wants bits {wants:#x}")
    ident = dockapp_id.encode("ascii")
    return (
        _header(KIND_HELLO)
        + struct.pack("<IBBBx", PROTOCOL_VERSION, tile_units, wants, len(ident))
        + token
        + ident
    )


def encode_frame(generation: int, width: int, height: int,
                 pixels: bytes) -> bytes:
    """``pixels`` is premultiplied RGBA8, top row first, no row padding."""
    if not (0 < width <= MAX_TILE_PX
            and 0 < height <= MAX_TILE_PX * MAX_TILE_UNITS):
        raise EncodeError(f"frame geometry {width}x{height} is out of range")
    expected = width * height * 4
    if len(pixels) != expected:
        raise EncodeError(
            f"frame needs {expected} pixel bytes but was given {len(pixels)}")
    if expected > MAX_FRAME_BYTES:
        raise EncodeError(f"frame of {expected} bytes exceeds the cap")
    return (
        _header(KIND_FRAME)
        + struct.pack("<III", generation & 0xFFFFFFFF, width, height)
        + pixels
    )


def encode_pong(seq: int) -> bytes:
    return _header(KIND_PONG) + struct.pack("<I", seq & 0xFFFFFFFF)


def encode_log(level: int, text: str) -> bytes:
    """Sanitized and truncated exactly as the reference encoder does:
    control characters (plus the Unicode line/paragraph separators, bidi
    controls and zero-width characters) are dropped, and the result is
    clipped to 256 bytes on a character boundary."""
    if level not in (LOG_ERROR, LOG_WARN, LOG_INFO, LOG_DEBUG):
        raise EncodeError(f"undefined log level {level}")
    out = []
    size = 0
    for c in text:
        cp = ord(c)
        dangerous = (
            cp < 0x20 or cp == 0x7F or 0x80 <= cp <= 0x9F
            or cp in (0x2028, 0x2029, 0x200B, 0x200C, 0x200D, 0xFEFF)
            or 0x202A <= cp <= 0x202E or 0x2066 <= cp <= 0x2069
        )
        if dangerous:
            continue
        b = c.encode("utf-8")
        if size + len(b) > MAX_LOG_BYTES:
            break
        out.append(c)
        size += len(b)
    encoded = "".join(out).encode("utf-8")
    return (
        _header(KIND_LOG)
        + struct.pack("<BxH", level, len(encoded))
        + encoded
    )


# ---------------------------------------------------------------------
# Shell -> client decoder
# ---------------------------------------------------------------------

def _decode_theme_state(body: bytes, kind_name: str) -> ThemeState:
    if len(body) < 16:
        raise DecodeError(f"{kind_name} ended inside its fixed fields")
    tile_px, scale_bits, id_len, reserved, toml_len = struct.unpack_from(
        "<IIHHI", body, 0)
    if reserved != 0:
        raise DecodeError(f"{kind_name} reserved field was not zero")
    scale = struct.unpack("<f", struct.pack("<I", scale_bits))[0]
    # The BadFloat rule: a scale a tile cannot be drawn at is rejected
    # here, in the codec, so the "resend when changed" loops above can
    # rely on a decoded state being equal to itself.
    if not (scale == scale and abs(scale) != float("inf")
            and 0.0 < scale <= MAX_SCALE):
        raise DecodeError(f"unusable scale (bits {scale_bits:#010x})")
    if id_len > MAX_THEME_ID_BYTES:
        raise DecodeError(f"theme_id of {id_len} bytes is over its cap")
    if toml_len > MAX_THEME_TOML_BYTES:
        raise DecodeError(f"theme_toml of {toml_len} bytes is over its cap")
    _exact(body, 16 + id_len + toml_len, kind_name)
    try:
        theme_id = body[16:16 + id_len].decode("utf-8")
        theme_toml = body[16 + id_len:16 + id_len + toml_len].decode("utf-8")
    except UnicodeDecodeError as e:
        raise DecodeError(f"{kind_name} carries invalid UTF-8") from e
    return ThemeState(tile_px=tile_px, scale=scale, theme_id=theme_id,
                      theme_toml=theme_toml)


def decode_server(buf: bytes):
    """Decodes one shell->client datagram.

    Returns a ``(name, payload)`` tuple, one of::

        ("welcome",       ThemeState)
        ("theme_changed", ThemeState)
        ("input",         InputEvent)
        ("visibility",    bool)
        ("ping",          seq: int)
        ("goodbye",       reason: int)   # GOODBYE_* codes

    Raises :class:`DecodeError` for anything else. A client's correct
    response to a DecodeError is to drop the connection — the two ends
    disagree about the protocol, and continuing would be guessing.
    """
    kind = _check_header(buf)
    body = buf[4:]
    if kind == KIND_WELCOME:
        return ("welcome", _decode_theme_state(body, "Welcome"))
    if kind == KIND_THEME_CHANGED:
        return ("theme_changed", _decode_theme_state(body, "ThemeChanged"))
    if kind == KIND_INPUT:
        _exact(body, 16, "Input")
        ev_kind, button, reserved, x, y, delta = struct.unpack("<BBHiii", body)
        if not INPUT_PRESS <= ev_kind <= INPUT_LEAVE:
            raise DecodeError(f"undefined input kind {ev_kind}")
        if button > BUTTON_RIGHT:
            raise DecodeError(f"undefined button {button}")
        if reserved != 0:
            raise DecodeError("Input reserved field was not zero")
        return ("input", InputEvent(ev_kind, button, x, y, delta))
    if kind == KIND_VISIBILITY:
        _exact(body, 4, "Visibility")
        if body[1:] != b"\x00\x00\x00":
            raise DecodeError("Visibility reserved bytes were not zero")
        if body[0] > 1:
            raise DecodeError(f"undefined visibility value {body[0]}")
        return ("visibility", body[0] == 1)
    if kind == KIND_PING:
        _exact(body, 4, "Ping")
        return ("ping", struct.unpack("<I", body)[0])
    if kind == KIND_GOODBYE:
        _exact(body, 4, "Goodbye")
        if body[1:] != b"\x00\x00\x00":
            raise DecodeError("Goodbye reserved bytes were not zero")
        if body[0] not in GOODBYE_NAMES:
            raise DecodeError(f"undefined goodbye reason {body[0]}")
        return ("goodbye", body[0])
    raise DecodeError(f"unknown message kind {kind:#04x}")

File: bindings/go/chonkdock/wire.go
// Package chonkdock implements the chonkstep dockapp protocol,
// version 1, in stdlib Go.
//
// A dockapp is a separate process that draws one (or a few) chonkstep
// dock tiles and pushes finished pixels to the desktop shell over a
// private SOCK_SEQPACKET Unix socket. It holds no display connection,
// crashes alone, restyles without restarting, and survives shell
// restarts. The wire contract this package implements is documented in
// docs/dockapp-protocol.md in the chonkstep repository; the normative
// implementation is crates/chonk-dock-proto.
//
// This file is the codec: pure, no I/O. Decoding is strict on purpose,
// mirroring the reference decoder — unknown kinds, non-zero reserved
// bytes, trailing bytes, out-of-range enums and unusable floats are
// rejected rather than clamped, because a client that silently
// mis-parses a message draws garbage with no error to explain it.
package chonkdock

import (
	"encoding/binary"
	"fmt"
	"math"
	"unicode/utf8"
)

// Protocol constants. See docs/dockapp-protocol.md section 3 for the
// reasoning behind each value.
const (
	ProtocolVersion   = 1
	TokenBytes        = 16
	MaxMessageBytes   = 256 * 1024
	MaxFrameBytes     = MaxMessageBytes - 64
	MaxTilePx         = 256
	MaxScale          = 8.0
	MaxTileUnits      = 4
	MaxIDBytes        = 64
	MaxLogBytes       = 256
	MaxThemeIDBytes   = 64
	MaxThemeTOMLBytes = 128 * 1024
)

// Message kinds. Client->shell in the low space, shell->client with
// the high bit set.
const (
	kindHello        = 0x01
	kindFrame        = 0x02
	kindPong         = 0x03
	kindLog          = 0x04
	kindWelcome      = 0x81
	kindThemeChanged = 0x82
	kindInput        = 0x83
	kindVisibility   = 0x84
	kindPing         = 0x85
	kindGoodbye      = 0x86
)

// Input-mask bits for Hello's wants field.
const (
	WantPress    = 1 << 0
	WantRelease  = 1 << 1
	WantScroll   = 1 << 2
	WantCrossing = 1 << 3
	WantAll      = WantPress | WantRelease | WantScroll | WantCrossing
)

// InputEvent kinds.
const (
	InputPress   = 1
	InputRelease = 2
	InputScroll  = 3
	InputEnter   = 4
	InputLeave   = 5
)

// InputEvent buttons (0 = none). Middle and Right exist on the wire
// but the shell never sends them: middle is the dock's reorder gesture
// and right opens the per-tile menu.
const (
	ButtonNone   = 0
	ButtonLeft   = 1
	ButtonMiddle = 2
	ButtonRight  = 3
)

// Log levels.
const (
	LogError = 1
	LogWarn  = 2
	LogInfo  = 3
	LogDebug = 4
)

// GoodbyeReason says why the shell is closing a connection.
type GoodbyeReason uint8

// Goodbye reasons.
const (
	GoodbyeShutdown      GoodbyeReason = 1
	GoodbyeProtocolError GoodbyeReason = 2
	GoodbyeUnauthorized  GoodbyeReason = 3
	GoodbyeReplaced      GoodbyeReason = 4
	GoodbyeTileTooLarge  GoodbyeReason = 5
	GoodbyeOverflow      GoodbyeReason = 6
	GoodbyeRemoved       GoodbyeReason = 7
)

func (r GoodbyeReason) String() string {
	switch r {
	case GoodbyeShutdown:
		return "Shutdown"
	case GoodbyeProtocolError:
		return "ProtocolError"
	case GoodbyeUnauthorized:
		return "Unauthorized"
	case GoodbyeReplaced:
		return "Replaced"
	case GoodbyeTileTooLarge:
		return "TileTooLarge"
	case GoodbyeOverflow:
		return "Overflow"
	case GoodbyeRemoved:
		return "Removed"
	default:
		return fmt.Sprintf("GoodbyeReason(%d)", uint8(r))
	}
}

// DecodeError means the peer's bytes could not be read as a v1
// message. A client's correct response is to drop the connection: the
// two ends disagree about the protocol, and continuing would be
// guessing.
type DecodeError struct{ What string }

func (e *DecodeError) Error() string { return "chonkdock: undecodable message: " + e.What }

func decodeErr(format string, args ...any) error {
	return &DecodeError{What: fmt.Sprintf(format, args...)}
}

// ThemeState is the geometry and palette a dockapp draws with, carried
// by Welcome and ThemeChanged. ThemeID is the fast path (look the
// palette up if you ship them); ThemeTOML is the correctness path (a
// serialized theme table, possibly empty). A client that can use
// neither falls back to its own default colors — wrong colors, never a
// blank tile.
type ThemeState struct {
	TilePx    uint32
	Scale     float32
	ThemeID   string
	ThemeTOML string
}

// InputEvent is one pointer event, in coordinates local to this
// dockapp's own tile.
type InputEvent struct {
	Kind   uint8 // Input* constants
	Button uint8 // Button* constants; ButtonNone for Scroll/Enter/Leave
	X, Y   int32
	Delta  int32 // scroll notches, signed; 0 for everything but Scroll
}

// ServerMessage is one decoded shell->client datagram. Exactly one of
// the payload fields is meaningful, selected by Kind.
type ServerMessage struct {
	Kind    string // "welcome" | "theme_changed" | "input" | "visibility" | "ping" | "goodbye"
	Theme   ThemeState
	Input   InputEvent
	Visible bool
	Seq     uint32
	Reason  GoodbyeReason
}

// IsValidID reports whether id satisfies the wire's id rule:
// 1..=64 bytes of [A-Za-z0-9._:-].
func IsValidID(id string) bool {
	if len(id) == 0 || len(id) > MaxIDBytes {
		return false
	}
	for i := 0; i < len(id); i++ {
		c := id[i]
		ok := c >= 'a' && c <= 'z' || c >= 'A' && c <= 'Z' ||
			c >= '0' && c <= '9' || c == '-' || c == '_' || c == '.' || c == ':'
		if !ok {
			return false
		}
	}
	return true
}

// FrameFits reports whether a tile of this geometry can be carried by
// the v1 inline transport. The shell refuses a Hello for which this is
// false, with Goodbye{TileTooLarge}.
func FrameFits(tilePx uint32, tileUnits uint8) bool {
	if tilePx == 0 || tilePx > MaxTilePx || tileUnits == 0 || tileUnits > MaxTileUnits {
		return false
	}
	return uint64(tilePx)*uint64(tilePx)*uint64(tileUnits)*4 <= MaxFrameBytes
}

func header(kind byte, capacity int) []byte {
	out := make([]byte, 4, 4+capacity)
	out[0] = kind
	return out
}

// EncodeHello builds the message a dockapp opens with. token is the 16
// raw bytes decoded from CHONKSTEP_DOCK_TOKEN.
func EncodeHello(id string, tileUnits uint8, token []byte, wants uint8) ([]byte, error) {
	if !IsValidID(id) {
		return nil, fmt.Errorf("chonkdock: invalid dockapp id %q", id)
	}
	if len(token) != TokenBytes {
		return nil, fmt.Errorf("chonkdock: token must be %d bytes, got %d", TokenBytes, len(token))
	}
	if wants&^uint8(WantAll) != 0 {
		return nil, fmt.Errorf("chonkdock: undefined wants bits %#x", wants)
	}
	out := header(kindHello, 24+len(id))
	out = binary.LittleEndian.AppendUint32(out, ProtocolVersion)
	out = append(out, tileUnits, wants, byte(len(id)), 0)
	out = append(out, token...)
	out = append(out, id...)
	return out, nil
}

// EncodeFrame builds one tile frame. pixels is premultiplied RGBA8,
// top row first, width*height*4 bytes with no row padding.
func EncodeFrame(generation, width, height uint32, pixels []byte) ([]byte, error) {
	if width == 0 || height == 0 || width > MaxTilePx || height > MaxTilePx*MaxTileUnits {
		return nil, fmt.Errorf("chonkdock: frame geometry %dx%d is out of range", width, height)
	}
	expected := int(width) * int(height) * 4
	if len(pixels) != expected {
		return nil, fmt.Errorf("chonkdock: frame needs %d pixel bytes, got %d", expected, len(pixels))
	}
	if expected > MaxFrameBytes {
		return nil, fmt.Errorf("chonkdock: frame of %d bytes exceeds the cap", expected)
	}
	out := header(kindFrame, 12+len(pixels))
	out = binary.LittleEndian.AppendUint32(out, generation)
	out = binary.LittleEndian.AppendUint32(out, width)
	out = binary.LittleEndian.AppendUint32(out, height)
	out = append(out, pixels...)
	return out, nil
}

// EncodePong answers a Ping, echoing its seq.
func EncodePong(seq uint32) []byte {
	return binary.LittleEndian.AppendUint32(header(kindPong, 4), seq)
}

// EncodeLog builds a diagnostic line for the session journal. The text
// is sanitized and truncated exactly as the reference encoder does:
// control characters, the Unicode line/paragraph separators, bidi
// controls and zero-width characters are dropped, and the result is
// clipped to 256 bytes on a rune boundary.
func EncodeLog(level uint8, text string) ([]byte, error) {
	if level < LogError || level > LogDebug {
		return nil, fmt.Errorf("chonkdock: undefined log level %d", level)
	}
	clean := make([]byte, 0, min(len(text), MaxLogBytes))
	for _, r := range text {
		dangerous := r < 0x20 || r == 0x7F || (r >= 0x80 && r <= 0x9F) ||
			r == 0x2028 || r == 0x2029 ||
			(r >= 0x202A && r <= 0x202E) || (r >= 0x2066 && r <= 0x2069) ||
			r == 0x200B || r == 0x200C || r == 0x200D || r == 0xFEFF
		if dangerous {
			continue
		}
		if len(clean)+utf8.RuneLen(r) > MaxLogBytes {
			break
		}
		clean = utf8.AppendRune(clean, r)
	}
	out := header(kindLog, 4+len(clean))
	out = append(out, level, 0)
	out = binary.LittleEndian.AppendUint16(out, uint16(len(clean)))
	out = append(out, clean...)
	return out, nil
}

func decodeThemeState(body []byte, what string) (ThemeState, error) {
	if len(body) < 16 {
		return ThemeState{}, decodeErr("%s ended inside its fixed fields", what)
	}
	tilePx := binary.LittleEndian.Uint32(body[0:4])
	scaleBits := binary.LittleEndian.Uint32(body[4:8])
	idLen := int(binary.LittleEndian.Uint16(body[8:10]))
	if body[10] != 0 || body[11] != 0 {
		return ThemeState{}, decodeErr("%s reserved field was not zero", what)
	}
	tomlLen := int(binary.LittleEndian.Uint32(body[12:16]))
	scale := math.Float32frombits(scaleBits)
	// The BadFloat rule: a scale a tile cannot be drawn at is rejected
	// in the codec, so a decoded state is always equal to itself and
	// "resend when changed" loops terminate.
	if math.IsNaN(float64(scale)) || math.IsInf(float64(scale), 0) || scale <= 0 || scale > MaxScale {
		return ThemeState{}, decodeErr("unusable scale (bits %#010x)", scaleBits)
	}
	if idLen > MaxThemeIDBytes {
		return ThemeState{}, decodeErr("theme_id of %d bytes is over its cap", idLen)
	}
	if tomlLen > MaxThemeTOMLBytes {
		return ThemeState{}, decodeErr("theme_toml of %d bytes is over its cap", tomlLen)
	}
	if len(body) != 16+idLen+tomlLen {
		return ThemeState{}, decodeErr("%s length %d disagrees with its declared strings", what, len(body))
	}
	id := body[16 : 16+idLen]
	toml := body[16+idLen:]
	if !utf8.Valid(id) || !utf8.Valid(toml) {
		return ThemeState{}, decodeErr("%s carries invalid UTF-8", what)
	}
	return ThemeState{TilePx: tilePx, Scale: scale, ThemeID: string(id), ThemeTOML: string(toml)}, nil
}

// DecodeServer decodes one shell->client datagram.
func DecodeServer(buf []byte) (ServerMessage, error) {
	var msg ServerMessage
	if len(buf) == 0 {
		return msg, decodeErr("empty message")
	}
	if len(buf) > MaxMessageBytes {
		return msg, decodeErr("message of %d bytes exceeds the cap", len(buf))
	}
	if len(buf) < 4 {
		return msg, decodeErr("message ended inside the header")
	}
	if buf[1] != 0 || buf[2] != 0 || buf[3] != 0 {
		return msg, decodeErr("reserved header bytes were not zero")
	}
	body := buf[4:]
	switch buf[0] {
	case kindWelcome, kindThemeChanged:
		theme, err := decodeThemeState(body, "Welcome/ThemeChanged")
		if err != nil {
			return msg, err
		}
		msg.Theme = theme
		if buf[0] == kindWelcome {
			msg.Kind = "welcome"
		} else {
			msg.Kind = "theme_changed"
		}
	case kindInput:
		if len(body) != 16 {
			return msg, decodeErr("Input is %d body bytes, want 16", len(body))
		}
		if body[0] < InputPress || body[0] > InputLeave {
			return msg, decodeErr("undefined input kind %d", body[0])
		}
		if body[1] > ButtonRight {
			return msg, decodeErr("undefined button %d", body[1])
		}
		if body[2] != 0 || body[3] != 0 {
			return msg, decodeErr("Input reserved field was not zero")
		}
		msg.Kind = "input"
		msg.Input = InputEvent{
			Kind:   body[0],
			Button: body[1],
			X:      int32(binary.LittleEndian.Uint32(body[4:8])),
			Y:      int32(binary.LittleEndian.Uint32(body[8:12])),
			Delta:  int32(binary.LittleEndian.Uint32(body[12:16])),
		}
	case kindVisibility:
		if len(body) != 4 {
			return msg, decodeErr("Visibility is %d body bytes, want 4", len(body))
		}
		if body[1] != 0 || body[2] != 0 || body[3] != 0 {
			return msg, decodeErr("Visibility reserved bytes were not zero")
		}
		if body[0] > 1 {
			return msg, decodeErr("undefined visibility value %d", body[0])
		}
		msg.Kind = "visibility"
		msg.Visible = body[0] == 1
	case kindPing:
		if len(body) != 4 {
			return msg, decodeErr("Ping is %d body bytes, want 4", len(body))
		}
		msg.Kind = "ping"
		msg.Seq = binary.LittleEndian.Uint32(body)
	case kindGoodbye:
		if len(body) != 4 {
			return msg, decodeErr("Goodbye is %d body bytes, want 4", len(body))
		}
		if body[1] != 0 || body[2] != 0 || body[3] != 0 {
			return msg, decodeErr("Goodbye reserved bytes were not zero")
		}
		if body[0] < uint8(GoodbyeShutdown) || body[0] > uint8(GoodbyeRemoved) {
			return msg, decodeErr("undefined goodbye reason %d", body[0])
		}
		msg.Kind = "goodbye"
		msg.Reason = GoodbyeReason(body[0])
	default:
		return msg, decodeErr("unknown message kind %#04x", buf[0])
	}
	return msg, nil
}
